src/zope/container/ordered.py

```diff
@@ -272,17 +273,14 @@ def updateOrder(self, order):
         if len(order) != len(self._order):
             raise ValueError("Incompatible key set.")
 
-        was_dict = {}
-        will_be_dict = {}
-        new_order = PersistentList()
+        order = [checkAndConvertName(x) for x in order]
 
-        for i, obj in enumerate(order):
-            was_dict[self._order[i]] = 1
-            will_be_dict[obj] = 1
-            new_order.append(obj)
-
-        if will_be_dict != was_dict:
+        if frozenset(order) != frozenset(self._order):
             raise ValueError("Incompatible key set.")
 
-        self._order = new_order
+        if order == self._order:
+            return
+
+        self._order.sort(key=order.index)
```

Comment:
> It's O(n log n) from what I understand about timsort, not quadratic

Normal timsort, sure, but the `order.index` key function makes it quadratic (I think).

WRT that test:  Identity isn't that meaningful a check for immutable "value" types.  And we've already done the allocation / initialization of the replacements, so we don't save anything there anyway.